docs/docs/concepts/durable_execution.md

```diff
@@ -0,0 +1,50 @@
+3. Wrap any non-deterministic operations (e.g., random number generation) or operations with side effects (e.g., file writes, API calls) inside [tasks][langgraph.func.task] to ensure that when a workflow is resumed, these operations are not repeated for the particular run, and instead their results are retrieved from the persistence layer. For more information, see [Determinism and Consistent Replay](#determinism-and-consistent-replay).
+
+## Determinism and Consistent Replay
+
+When you resume a workflow run, the code does **NOT** resume from the **same line of code** where execution stopped; instead, it will identify an appropriate [starting point](#starting-points-for-resuming-workflows-advanced) from which to pick up where it left off. This means that the workflow will replay all steps from the [starting point](#starting-points-for-resuming-workflows-advanced) until it reaches the point where it was stopped.
+
+As a result, when you're writing a workflow for durable execution, you should wrap any non-deterministic operations (e.g., random number generation) and any operations with side effects (e.g., file writes, API calls) inside [tasks](./functional_api.md#task).
+
+- **Avoid Repeating Work**  Place operations that produce side effects (such as logging, file writes, or network calls) after an interrupt in Graph API nodes or encapsulate them within [tasks](./functional_api.md#task). This prevents their unintended repetition when the workflow is resumed. LangGraph will look up information stored in the persistence layer about any **tasks** that were previously executed for the specific run and will swap in the results of those tasks instead of re-executing them.
+- **Encapsulate Non-Deterministic Operations:**  Wrap any code that might yield non-deterministic results (e.g., random number generation) inside **tasks**. This ensures that, upon resumption, the workflow follows the exact recorded sequence of steps with the same outcomes.
+
+For some examples of pitfalls to avoid, see the [Common Pitfalls](./functional_api.md#common-pitfalls) section in the functional API, which shows
+how to structure your code using **tasks** to avoid these issues. The same principles apply to the Graph API (you can use **tasks** from inside [nodes](./low_level.md#nodes) as well).
+
+## Idempotency
+
+Idempotency ensures that running the same operation multiple times produces the same result. This helps prevent duplicate API calls and redundant processing if a step is rerun due to a failure. Always place API calls inside **tasks** functions for checkpointing, and design them to be idempotent in case of re-execution. Re-execution can occur if a **task** starts, but does not complete successfully. Then, if the workflow is resumed, the **task** will run again. Use idempotency keys or verify existing results to avoid duplication.
```

Comment:
I think it would be clearer to mention idempotency in the section above. Making side-effect-producing logic idempotent should be the first suggestion, IMO.

Reply:
Agreed, merged into the previous paragraph. It's the third bullet point rather than the first, since the main objective of the conceptual doc is to the happy path of durable execution (i.e., what it does and how to set it up).